### DISCUSSION TITLE
How should mdx-js/loader v2 works with CRA?

### DISCUSSION THREAD

**Discussion by askides** [mdx-js/mdx#1870] [Q&A] [answered]
Hello,

I'm having a problem using MDX with create-react-app, i tried i every possible way for hours but without success..

The problem is this: following the [documentation](https://mdxjs.com/docs/getting-started/#create-react-app-cra) i installed the @mdx-js/loader, and as the docs say: 'There is no need to configure it'.

But when i create the example file, (i tried with the same example from the docs, but also other examples), React throws an exception telling me something like this:

```
failed to execute 'createElement' on 'document':
```

With the debugger i can see that React is trying to call createElement with the mdx file path as a parameter.

```
createElement('./static/29fh30fh20dh40e.mdx');
```

I noticed also, that the result does not change if i omit the inline babel import of mdx-js/loader. The error is the same:

```
import Content from "!@mdx-js/loader!./content.mdx";
import Content from "./content.mdx";
```

There are the version i'm using for these experiments:

```
"@mdx-js/loader": "^2.0.0-rc.2"
"@mdx-js/mdx": "^2.0.0-rc.2"
"react": "^17.0.2"
"react-dom": "^17.0.2"
```
    
PS: After some trials, i decided to try it also on Next.js, and following the docs, it works perfectly, so the problem should be only in CRA.

Thank you guys.

**Comment by wooorm** [comment]
Could definitely be that CRA is not supported currently. Did you try their latest, v5?

**Reply by askides** [reply]
Yes, i tried with the last one CRA.

Also installed a brand new one with npx.

**Comment by seyoon20087** [comment]
I'm also experiencing the same issue (copy of https://github.com/mdx-js/mdx/issues/1871#issue-1088328732).

package.json:
```json
{
  "private": true,
  "dependencies": {
    "@mdx-js/loader": "^2.0.0-rc.2",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-scripts": "5.0.0",
    "web-vitals": "^2.1.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

src/content.mdx:
```mdx
# Hello, world!

This is **markdown** with <span style={{color: "red"}}>JSX</span>: MDX!
```

src/App.js:
```jsx
/* eslint-disable import/no-webpack-loader-syntax */
import Content from '!@mdx-js/loader!./content.mdx'

export default function App() {
  return <Content />
}
```

If I try to do it an error is thrown in the console, something like this:

```console
Uncaught DOMException: Failed to execute 'createElement' on 'Document': The tag name provided ('/static/media/content.152fde8da01171ae4224.mdx') is not a valid name.
    at createElement (http://localhost:3000/static/js/bundle.js:16306:38)
    at createInstance (http://localhost:3000/static/js/bundle.js:17493:24)
    at completeWork (http://localhost:3000/static/js/bundle.js:26619:32)
    at completeUnitOfWork (http://localhost:3000/static/js/bundle.js:29924:20)
    at performUnitOfWork (http://localhost:3000/static/js/bundle.js:29896:9)
    at workLoopSync (http://localhost:3000/static/js/bundle.js:29822:9)
    at renderRootSync (http://localhost:3000/static/js/bundle.js:29788:11)
    at performSyncWorkOnRoot (http://localhost:3000/static/js/bundle.js:29405:22)
    at scheduleUpdateOnFiber (http://localhost:3000/static/js/bundle.js:28993:11)
    at updateContainer (http://localhost:3000/static/js/bundle.js:32530:7)
```

**Reply by pannchat** [reply]
I'm experiencing the same issue. 🥲

**Reply by pannchat** [reply]
I think you can lower the @mdx-js/loader version in the CRA environment.
and the react-scripts also version is 4.0.3.

**Comment by 7i7o** [comment]
Hi! I'm experiencing the same exact issue on a fresh CRA. I wanted to know if this has been fixed or someone found a workaround, since i'm following the getting started guides literally without success.

**Comment by jamescho7** [comment]
I was able to get a working configuration by downgrading @mdx-js/loader to 1.6.22 and react-scripts to 4.0.3.

Delete node_modules and package-lock.json in your project, update your dependency versions in package.json to:

```
"@mdx-js/loader": "^1.6.22",
"react-scripts": "4.0.3",
```
Then run npm install to re-install your dependencies.

When importing .mdx files using mdx-js/loader, you may get an error parsing the resulting JSX data. I was able to resolve this by using babel-loader as well

```
/* eslint-disable import/no-webpack-loader-syntax */
import Content from "!babel-loader!!@mdx-js/loader!./documents/sample.mdx";
```

You may need to create a .babelrc file in your root directory with the following presets:

`{ "presets": ["@babel/preset-env", "@babel/preset-react"] }`

**Reply by Steven19Lee** [reply]
Hi! Thanks for your solution, It works in my local environment, but in code sandbox, it still fails, is it convenient if you take a look at my code sandbox there? link(https://codesandbox.io/s/mdx-test-su6bzy?file=/.babelrc.json). thanks for any help!

**Comment by julienlaurent-migso** [comment]
Hi, 

I am facing the same issue, but i don't want to downgrade my libraries to prevent dependencies vulnerabilties. 

There is a solution to use mdx/loader with create-react-app like describe in the mdx documentation ?

how i can compile my .mdx file to execute 'createElement' like <MyMdxFil /> ?

Than you for your help,

**Comment by BastienFerrier** [comment]
Hi,

I'm experiencing the same problem, does someone found out any fix or workaround for it?

Thanks'

**Comment by brian-petersen** [comment]
I haven't yet found a fix for this, but I want to document some of my findings:

1. This is a problem with the `create-react-app` update to 5.0.0
2. The specific problem is in their upgrade to Webpack 5 as a part of the `create-react-app` update
3. The problem is that the asset module (previously `file-loader`) loader is being applied
4. Because of that, a path to the file (and not the generated react component) is imported

Relevant links:

1. https://github.com/facebook/create-react-app/pull/11201/files?diff=split&w=0#diff-8e25c4f6f592c1fcfc38f0d43d62cbd68399f44f494c1b60f0cf9ccd7344d697R573

I confirmed this by changing this line locally and running `npm run start`: https://github.com/facebook/create-react-app/blob/fd8c5f7b1b1d19d10d24cc2f9fdfc110585dc030/packages/react-scripts/config/webpack.config.js#L595 from

```
exclude: [/^$/, /\.(js|mjs|jsx|ts|tsx)$/, /\.html$/, /\.json$/],
```

to

```
exclude: [/^$/, /\.(js|mjs|jsx|ts|tsx|mdx)$/, /\.html$/, /\.json$/],
```

Once I did that, the file contents were picked up as I expected.

I'm not sure if this is a bug in MDX, CRA, or Webpack.

As I understand it, using the syntax `import Content from "!!@mdx-js/loader!./views/terms/PrivacyPolicy/Content.mdx"` should bypass all other loaders (as documented in https://webpack.js.org/concepts/loaders/) -- but that's not working.

**Reply by wooorm** [reply]
Given that you can change CRA to make MDX work, that seems to indicate that CRA can be patched to make MDX work, so perhaps you can ask CRA folks to apply that patch?

**Reply by wzard** [reply]
This solution worked for me. Using CRA5 with the latest @mds-js/loader. 
Folks using react-app-rewired can use the following to exclude mdx files 

```js
   config.module.rules[1].oneOf = config.module.rules[1].oneOf.map((one) => {

      
      if (one.exclude) {
        one.exclude = [/^$/, /\.(js|mjs|jsx|ts|tsx|mdx)$/, /\.html$/, /\.json$/]
      }
      return one;
    })
    return config;
```

**Reply by haendzel** [reply]
@wzard Where? What file to edit?

**Reply by wzard** [reply]
@haendzel If you install [react-app-rewired](https://github.com/timarney/react-app-rewired), then you can override webpack configuration by writing a `config-override.js` file. In that, you can put the above code.

**Reply by risu-p** [reply]
```
config.module.rules[1].oneOf = config.module.rules[1].oneOf.map((one) => {
  if (one.exclude && one.exclude instanceof Array) {
    one.exclude.push(/\.mdx?$/);
  }
  return one;
});
```

**Comment by mtajammulzia** [comment]
Any updates on this one? I have been trying to configure React with MDX for two days now but nothing seems to work.

I have tried using react-app-rewired and config-overrides.js but it fails and gives `getBabelLoader() is deprecated` error.

Following thread.

**Comment by zingbretsen** [comment]
I opened an issue in CRA here:
https://github.com/facebook/create-react-app/issues/12166

The fix from @brian-petersen above (https://github.com/mdx-js/mdx/discussions/1870#discussioncomment-2304871) works for me locally.

**Reply by zingbretsen** [reply]
and PR opened here: https://github.com/facebook/create-react-app/pull/12167

**Comment by aakanshaverma11** [comment]
Hey @zingbretsen did you receive any update on merging the file ? This was really help, it works fine my local but not through pipeline.

**Reply by zingbretsen** [reply]
No updates yet. I switched to NextJS for the project I was working on 🤷

**Comment by userzimmermann** [comment]
Thanks @brian-petersen – After hours of frustration your https://github.com/mdx-js/mdx/discussions/1870#discussioncomment-2304871 saved my day … OMG: late evening already :D 

This discussion is unfortunately hard to find when searching for `Uncaught DOMException: Failed to execute 'createElement' on 'Document': The tag name provided ...`

Anyway: Not only `mdx` should be added to the list, but also plain `md` – Please add the latter to your https://github.com/facebook/create-react-app/pull/12167 PR as well, @zingbretsen

**Comment by mtajammulzia** [comment]
Can anyone help me out? I have the similar problem but @brian-petersen method doesn't fix it for me.

I have added 

````js
{
        test: /\.mdx?$/,
        use: [
          {
            loader: '@mdx-js/loader',
            /** @type {import('@mdx-js/loader').Options} */
            options: {}
          }
        ]
      }
````
in the rules of webpack.config.js and also changed the line mentioned in his comment and added "|mdx"

I have installed @mdx-js/loader, @types/mdx and @mdx-js/react (not relevant)

Any ideas what might be wrong?

**Reply by userzimmermann** [reply]
Should be `loader: require.resolve('@mdx-js/loader')`, @mtajammulzia 

Have you tried it w/o your `test: ..., use: [{ loader: ...` definition; by just using explicit `import ... from '!@mdx-js/loader!...'`? I'm going that latter way; but your addition looks quite convenient … I'll also give it a try :)

And please post your error output. It's hard to help w/o ;)

**Reply by userzimmermann** [reply]
It works :) —by defining the loader simply like this (above the `// "file" loader ...` entry):

```js
            {
              test: /\.(md|mdx)$/,
              loader: require.resolve('@mdx-js/loader')
            }
```

Thanks for the idea, @mtajammulzia! & @zingbretsen: You should also add this to your https://github.com/facebook/create-react-app/pull/12167 PR

**Comment by userzimmermann** [comment]
Finally: I've found an elegant workaround :) —using https://github.com/gsoft-inc/craco & the following `craco.config.js`

```js
const { addAfterLoader, loaderByName } = require('@craco/craco')

module.exports = {
    webpack: {
        configure: (webpackConfig) => {
            addAfterLoader(webpackConfig, loaderByName('babel-loader'), {
                test: /\.(md|mdx)$/,
                loader: require.resolve('@mdx-js/loader')
            })

            return webpackConfig
        }
    }
}
```

**Reply by thediveo** [reply]
Unfortunately, I can't get this to work.
- upgraded CRA to 5.0.0, because that's the raison être.
- using https://www.npmjs.com/package/mdx-loader as a wrapper -- because I still cannot figure out how to configure @mdx-js/loader, because the documentation is right hostile to newcomers in this respect ... please help!
- added `craco.config.js` as shown above.
- yarn start

Trying to load the corresponding MDX element via lazy fails with `Uncaught Error: Element type is invalid. Received a promise that resolves to: /static/media/Lxkns.cbf76125c98b291b503c.mdx. Lazy element type must resolve to a class or function.`

**Reply by userzimmermann** [reply]
https://www.npmjs.com/package/mdx-loader is horribly outdated ;) – Just `yarn add @mdx-js/loader` – No further configuration needed … https://mdxjs.com/docs/getting-started/#create-react-app-cra is basically right about this …

And regarding CRACO: Have you adapted your `"scripts"` in `package.json` according to https://github.com/gsoft-inc/craco/blob/master/packages/craco/README.md#installation ?

**Reply by thediveo** [reply]
looks like I need to repeat myself as the important point got lost and hasn't been answered: how do I configure to use the plugins, add-ons, transformers or whatever they're called this second in order to get the additional value offered by mdx-loader but now without mdx-loader? Where do I place this and how does it look like? Does this always require craco?

**Reply by wooorm** [reply]
You never asked that and your terminology is all over the place. Please see https://mdxjs.com/community/support/ on how to ask questions. In particular, you’ll likely find the docs on how to use MDX useful: https://mdxjs.com/docs/using-mdx/

**Reply by userzimmermann** [reply]
> Does this always require craco?

As long as https://github.com/facebook/create-react-app/issues/12166 is not properly fixed, it always requires some _rewiring:_ that term made popular by the original https://github.com/timarney/react-app-rewired … CRACO is just currently—AFAIK—the only one of those tools that is _compatible enough_ with CRA 5

And when using CRACO, the setup steps are:

1. Install libraries:
```
$ yarn add @mdx-js/loader
$ yarn add @craco/craco
```

2. Create `./craco.config.js` according to https://github.com/mdx-js/mdx/discussions/1870#discussioncomment-2530611
3. Replace `react-scripts` with `craco` in your `project.json`'s `"scripts"` section
 
Then everything is ready to `yarn start` your app …

**Reply by thediveo** [reply]
@userzimmermann thank you very much for clarifying that part!

@wooorm Oh yes, I asked, please see my bullet points in this thread. And yes, my terminology is bad, as I'm no web stack dev by trade. If that causes you to disregard my problems, then this is exactly my point in case that the documentation is hostile to non-experts as it expects everyone to be perfect. So again, how can I replicate the functionality so far offered by mdx-loader using @mdx-js/loader, especially where do I the "mdx translation" configuration before done by mdx-loader? Is this completely non-understandable even if I'm not a perfect web dev?

**Reply by userzimmermann** [reply]
We're just trying to help you, @thediveo <3

**Reply by thediveo** [reply]
@userzimmermann you do, thank you again!

**Reply by userzimmermann** [reply]
You're welcome, @thediveo —& _we all_ know how nerve-wracking full-stack web development can sometimes be ;) 

You should nevertheless take a look at those links given by @wooorm —because they're indeed very useful as well!

**Reply by thediveo** [reply]
I had extensively looked at the MDX documentation, not only once but multiple times; yedt, I still don't get all the pieces working together ... that's also a reason for using `mdx-loader` so far. Anyway.

I installed `yarn add @mdx-js/loader` and then tried several different import version, which all fail with the same error message in the browser console I already mentioned above: `Uncaught Error: Element type is invalid. Received a promise that resolves to: /static/media/Lxkns.c0153dde72127f40ad39.mdx. Lazy element type must resolve to a class or function.` I tried these variations without any success:

```typescript
const ch = (name: string) => React.lazy(() => import(`!@mdx-js/loader!./chapters/${name}.mdx`))
```

```typescript
const ch = (name: string) => React.lazy(() => import(`./chapters/${name}.mdx`))
```

```typescript
const ch = (name: string) => React.lazy(() => import(`!babel-loader!@mdx-js/loader!./chapters/${name}.mdx`))
```

```typescript
const ch = (name: string) => React.lazy(() => import(`!babel-loader!./chapters/${name}.mdx`))
```

What am I doing wrong?

**Reply by userzimmermann** [reply]
Please post the contents of your `package.json`'s `"scripts: {...}"` property, @thediveo

**Reply by thediveo** [reply]
`package.json` with `scripts` property and `babel` property in case that should play a role (which I don't know):
```json
{
  "scripts": {
    "start": "REACT_APP_GIT_VERSION=${GIT_VERSION:=$(git describe)} PORT=3030 react-scripts start",
    "build": "REACT_APP_GIT_VERSION=${GIT_VERSION:=$(git describe)} react-scripts build",
    "build:dev": "REACT_APP_GIT_VERSION=${GIT_VERSION:=$(git describe)} react-scripts build --profile",
    "imagebuild": "REACT_APP_GIT_VERSION=$GIT_VERSION react-scripts build",
    "icons": "node genicons",
    "cypress": "NODE_ENV=test cypress open",
    "cypress:run": "NODE_ENV=test cypress run",
    "eject": "react-scripts eject",
    "styleguide": "REACT_APP_GIT_VERSION=$(git describe) npx styleguidist server",
    "styleguide:build": "REACT_APP_GIT_VERSION=$(git describe) npx styleguidist build"
  },
  "babel": {
    "presets": [
      "babel-preset-react-app"
    ],
    "plugins": [
      [
        "@babel/plugin-proposal-private-property-in-object",
        {
          "loose": true
        }
      ]
    ]
  },
}
```

**Reply by userzimmermann** [reply]
In https://github.com/mdx-js/mdx/discussions/1870#discussioncomment-2535641 I told you that `react-scripts` has to be replaced with `craco` in your `"scripts"` entries … Because otherwise, CRACO is not used—which leads exactly to your errors!

The same is explained in https://github.com/gsoft-inc/craco/blob/master/packages/craco/README.md#installation —which I had mentioned in https://github.com/mdx-js/mdx/discussions/1870#discussioncomment-2535052 before …

When you don't follow the instructions, then of course things will not work ;)

**Reply by thediveo** [reply]
Yes, absolutely my fault ... so thank you for spotting this and pointing it out to me ("Betriebsblindheit" 😬). I totally missed that important point (even when rereading), as I was somehow thinking that CRACO is a transparent drop-in add-on. Of course, this is technically impossible, so I should have noticed that myself.

The basic integration now works and interestingly brings up issues in my source files that weren't reported so far. But that's in my ballpark anyway.

Still, my other topic where I don't understand after reading how this is supposed to work, maybe that's just a tiny thing which I'm again missing and overlooking all the time. For instance, the stale `mdx-loader` "pulls in" `remark-emoji`, `remark-images`, ... Now, https://mdxjs.com/docs/extending-mdx/ lists "Options passed to the compiler" which doesn't seem to apply in my case, as the API use is abstracted away by using `@mdx-js/loader`, correct? Next bullet point is "Plugins [...] (see remark plugins [...])". Now, this document page further talks about how to use the API, if I'm not mistaken by looking at the code examples. Then, "Creating plugins" isn't suitable for me either.

When looking at https://mdxjs.com/packages/loader/ it tells me to configure via `webpack.config.js` ... which I never had when using CRA, so this always send me into a dead end. Now, with CRACO I suspect that I can add this to `craco.config.js`, but there is a difference between your fix for CRA, where you specify the `loader` property as a sibling to `test`, whereas https://mdxjs.com/packages/loader/#use uses a `use` property of type array with `loader` and `options`. What would be the correct form?

**Reply by wooorm** [reply]
>  please see my bullet points in this thread

Please do read it again. It does not include “how do I configure to use the plugins [...] in order to get the additional value offered by `mdx-loader`”. It says: “I still cannot figure out how to configure `@mdx-js/loader`”.
That is a different question. Something for another discussion. Something which you should provide much more context for. And the support docs guide you towards how to solve your problems and how to ask questions. They prevent this going back-and-forth.

---

@userzimmermann Thanks for helping and the wonderful insights.
@thediveo Please follow the support guide and ask questions in the appropriate venues according to our support docs or I’ll block you.

**Reply by padunk** [reply]
![Screen Shot 2022-06-17 at 16 37 26](https://user-images.githubusercontent.com/25411350/174272378-5b4f11ab-3482-43f0-92c3-d30d0b7517fd.png)
Hi @userzimmermann ,
Thanks for uploading the workaround, but I found this error when starting my apps.
Do you have any idea why is this happening?
Thanks again.

**Reply by wooorm** [reply]
Your webpack is old. Update it

**Reply by padunk** [reply]
I'm using webpack 5.73.0

**Reply by wooorm** [reply]
Webpack 5 has it: https://webpack.js.org/api/loaders/#thisgetoptionsschema.
Maybe you use an old CRA which uses an old webpack, different from your own.
It is impossible to help you with so little information.

**Reply by Jerrylum** [reply]
Can confirm this works with CRA 5.0.1 and `@mdx-js/loader 2.3.0`. Good job 👍

**Comment by ChristianMurphy** [comment, marked as answer]
Summarizing the current status.
Create React App is deprecated and is no longer actively being maintained. https://github.com/reactjs/react.dev/pull/5487#issuecomment-1409720741

The best course of action is to migrate to a build tool which is maintained.
Vite and Vitest have similar low configuration setup and extensive features set as Create React App had.
Some resources showing how to migrate:
* https://cathalmacdonnacha.com/migrating-from-create-react-app-cra-to-vite
* https://cathalmacdonnacha.com/migrating-from-jest-to-vitest
* https://dev.to/henriquejensen/migrating-from-create-react-app-to-vite-a-quick-and-easy-guide-5e72
* https://dev.to/dmnchzl/migrate-a-webapp-from-cra-to-vite-ebe
* https://medium.com/@guitartsword/how-to-migrate-from-create-react-app-to-vite-348dbdcfd894

You can find documentation on how MDX integrates with vite here: https://mdxjs.com/docs/getting-started/#vite

If you'd like to add server side rendering support while migrating, React has some recommendations https://react.dev/learn/start-a-new-react-project (all of which also have MDX integrations)

---

If you are unable to migrate to a actively maintained build tool, see https://github.com/orgs/mdx-js/discussions/1870#discussioncomment-2530611 for a workaround for Create React App using CRACO.
